src/adr/0022-cassandra-message-inconsistency.md

```diff
@@ -0,0 +1,88 @@
+# 22. Cassandra Message inconsistencies
+
+Date: 2020-02-27
+
+## Status
+
+Proposed
+
+## Context
+
+Messages are denormalized in Cassandra in order to:
+ - access them by their unique identifier (messageId), for example through the JMAP protocol
+ - access them by their mailbox identifier and Unique IDentifier within that mailbox (mailboxId + uid), for example 
+ through the IMAP protocol
+
+ - `messageIdTable` Holds mailbox and flags for each message, lookup by mailbox ID + UID
+ - `imapUidTable` Holds mailbox and flags for each message, lookup by message ID
+
+Failures during the denormalization process will lead to inconsistencies between the two tables.
+
+This can lead to the following user experience:
+
+```
+BOB receives a message
+The denormalization process fails
+BOB can read the message via JMAP
+BOB cannot read the message via IMAP
+
+BOB marks a message as SEEN
+The denormalization process fails
+The message is SEEN in JMAP
+The message is UNSEEN in IMAP
+```
+
+### Current operations
+
+ - Adding a message:
+   - (CassandraMessageMapper) First reference the message in `messageIdTable` then in `imapUidTable`.
+   - (CassandraMessageIdMapper) First reference the message in `imapUidTable` then in `messageIdTable`.
+ - Deleting a message:
+   - (CassandraMessageMapper) First delete the message in `imapUidTable` then in `messageIdTable`.
+   - (CassandraMessageIdMapper) Read the message metadata using `imapUidTable`, then first delete the message in 
+   `imapUidTable` then in `messageIdTable`.
+ - Copying a message:
+   - (CassandraMessageMapper) Read the message first, then first reference the message in `messageIdTable` then
+    in `imapUidTable`.
+ - Moving a message:
+   - (CassandraMessageMapper) Logically copy then delete. A failure in the chain migh lead to duplicated message (present 
+   in both source and destination mailbox) as well as different view in IMAP/JMAP.
+   - (CassandraMessageIdMapper) First reference the message in `imapUidTable` then in `messageIdTable`.
+ - Updating a message flags:
+   - (CassandraMessageMapper) First update conditionally the message in `imapUidTable` then in `messageIdTable`.
+   - (CassandraMessageIdMapper) First update conditionally the message in `imapUidTable` then in `messageIdTable`.
+
+## Decision
+
+Adopt `imapUidTable` as a source of truth. All writes will be performed to `messageIdTable` then performed on
```

Comment:
why not having taken `messageIdTable` as source of truth?

I think the explanation need to be in this ADR

Reply:
(for me the advantage of messageIdTable is that it can be a source of truth for copying/moving messages)